test/correctness/simd_op_check.h

```diff
@@ -0,0 +1,301 @@
+                        // signed integer overflow.
+                        *ptr = ((uint32_t)rng()) & 0x0fffffff;
+                    }
+                }
+
+            }
+            Realization r = error.realize();
+            double e = Buffer<double>(r[0])();
+            // Use a very loose tolerance for floating point tests. The
+            // kinds of bugs we're looking for are codegen bugs that
+            // return the wrong value entirely, not floating point
+            // accuracy differences between vectors and scalars.
+            if (e > 0.001) {
+                error_msg << "The vector and scalar versions of " << name << " disagree. Maximum error: " << e << "\n";
+
+                string error_filename = output_directory + "error_" + name + ".s";
+                error.compile_to_assembly(error_filename, arg_types, target);
+
+                std::ifstream error_file;
+                error_file.open(error_filename);
+
+                error_msg << "Error assembly: \n";
+                string line;
+                while (getline(error_file, line)) {
+                    error_msg << line << "\n";
+                }
+
+                error_file.close();
+            }
+        }
+
+        return { op, error_msg.str() };
+    }
+
+    void check(string op, int vector_width, Expr e) {
+        // Make a name for the test by uniquing then sanitizing the op name
+        string name = "op_" + op;
+        for (size_t i = 0; i < name.size(); i++) {
+            if (!isalnum(name[i])) name[i] = '_';
+        }
+
+        name += "_" + std::to_string(tasks.size());
+
+        // Bail out after generating the unique_name, so that names are
+        // unique across different processes and don't depend on filter
+        // settings.
+        if (!wildcard_match(filter, op)) return;
+
+        tasks.emplace_back(Task {op, name, vector_width, e});
+    }
+    virtual void add_tests() = 0;
+    virtual void setup_images() {
+        for (auto p : image_params) {
+            p.reset();
+            p.set_host_alignment(128);
+        }
+    }
+    virtual bool test_all() {
+        /* First add some tests based on the target */
+        add_tests();
+        Halide::Internal::ThreadPool<TestResult> pool(num_threads);
+        std::vector<std::future<TestResult>> futures;
+        for (const Task &task : tasks) {
+            futures.push_back(pool.async([this, task]() {
+                return check_one(task.op, task.name, task.vector_width, task.expr);
+            }));
+        }
+
+        bool success = true;
+        for (auto &f : futures) {
+            const TestResult &result = f.get();
+            std::cout << result.op << "\n";
+            if (!result.error_msg.empty()) {
+                std::cerr << result.error_msg;
+                success = false;
+            }
+        }
+
+        return success;
+    }
+};
+
+    //}  // namespace
+
+
+#endif
```

Comment:
It's good form to have a comment indicating the matching ifdef , ie `#endif  // SIMD_OP_CHECK_H`